### PR TITLE
feat(subagent): add redact_secrets option for context isolation

### PR DESCRIPTION
**Pull request by TimeToBuildBob**
## Summary

Fixes #2949 — subagent context isolation: secrets in workspace context.

Key clarification from code review: **subagents do NOT inherit parent conversation history** — they start fresh. What they *do* inherit (in `context_mode="full"`) is workspace context: files listed in `gptme.toml [prompt] files` and `context_cmd` output. If those contain secrets (API keys, tokens, passwords), they reach the subagent.

### Changes

- **New `redact_secrets=True` parameter** on `subagent()`: scrubs common secret patterns from workspace context messages before the subagent sees them. Patterns: variable names matching `API_KEY`, `TOKEN`, `PASSWORD`, `PRIVATE_KEY`, `AUTH_KEY`, `ACCESS_KEY`, `CREDENTIAL`. The key name and separator are preserved; only the value becomes `[REDACTED]`.
- **New `gptme/tools/subagent/context.py`**: `redact_secrets_from_text()` and `redact_secrets_from_messages()` utilities.
- **Updated `instructions` string**: adds a "Context Isolation" section documenting exactly what subagents do and don't inherit — so users understand the model without reading source code.
- **15 new unit tests** covering redaction of API keys, tokens, passwords, export statements, multi-line content, and `redact_secrets_from_messages()`.

### Usage

```python
# Sanitize workspace context before passing to subagent
subagent("analyzer", "Analyze the codebase", redact_secrets=True)

# For stronger isolation (no workspace files at all):
subagent("analyzer", "...", context_mode="selective", context_include=["agent"])
```

### What is and isn't fixed

| Concern from issue | Status |
|---|---|
| "full history" leaked to subagent | Already not true — subagents start fresh |
| Secrets in workspace config files/context_cmd | Fixed by `redact_secrets=True` |
| Workspace boundary ambiguity | Documented in instructions string |
| Subprocess/ACP context isolation | Subprocess already runs as separate process; no change needed |

## Test plan

- [x] `uv run python -m pytest tests/test_subagent_unit.py` — all 88 tests pass
- [x] New tests: `TestRedactSecretsFromText` (10 cases) + `TestRedactSecretsFromMessages` (5 cases)
- [x] Pre-commit hooks pass (ruff, mypy)

### CODE REVIEW

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Addressed the Greptile 3/5 findings in 4bd988a:

**P1 — `_SECRET_LINE_RE` dead code / YAML secrets unredacted**: Added `_COLON_ASSIGN_RE` (a pattern analogous to `_ENV_ASSIGN_RE` but using `:` as separator) and wired it as a fallback in `_redact_line()`. YAML/TOML-style lines like `github_token: ghp_xyz` and `openai_api_key: sk-proj-abc` are now correctly redacted.

**P2 — Unconditional `\n` appended**: `_redact_line` now preserves the original line ending — if the input line didn't end with `\n` (e.g. the last line of a file), the redacted output won't add one.

**P2 — `redact_secrets` silently ignored in subprocess/ACP modes**: Added a `logger.warning` when `redact_secrets=True` is passed for a non-thread-mode subagent, so misconfigured callers get a runtime signal rather than silent no-op.

**Tests**: 7 new tests in `TestRedactSecretsColonStyle` covering YAML-style keys, indented keys, multiline content, non-secret key preservation, and the trailing-newline edge case. All 22 redaction tests pass.

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/subagent/context.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+            .*?                              # anything after, before separator
+            (?:\s*[=:]\s*)?                  # optional repeated separator
+        )?
+    )
+    (["']?)                                  # group 2: optional opening quote
+    (\S+)                                    # group 3: the secret value
+    (["']?)                                  # group 4: optional closing quote
+    """,
+    re.MULTILINE,
+)
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py]: Fixed in 4bd988a: added `_COLON_ASSIGN_RE` for YAML/TOML-style colon assignments, wired as fallback in `_redact_line()`. Dead `_SECRET_LINE_RE` was removed in 72989fef. YAML-style secrets like `github_token: ghp_xyz` and `openai_api_key: sk-proj-abc` are now correctly redacted.

File: gptme/tools/subagent/context.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    # Try the env-var assignment pattern first (more specific)
+    match = _ENV_ASSIGN_RE.search(line)
+    if match:
+        export_prefix = match.group(1) or ""
+        name = match.group(2)
+        sep = match.group(3)
+        trailing = match.group(7)
+        # Reconstruct with redacted value (no quotes around [REDACTED])
+        return f"{export_prefix}{name}{sep}{_REDACTED}{trailing}\n"
+    return line
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py:120]: Fixed in 4bd988a: `_redact_line` now captures the original line ending and preserves it — `ending = "\n" if line.endswith("\n") else ""` — so the last line of a file without a trailing newline is not altered in length.

File: gptme/tools/subagent/api.py [hunk 3]
```diff
@@ -113,6 +114,20 @@ def subagent(
...
+            context messages before they are passed to the subagent.
+            Redacts values from lines where the variable name matches patterns
+            like API_KEY, TOKEN, PASSWORD, PRIVATE_KEY, etc.
+
+            Note: subagents do NOT inherit the parent's conversation history —
+            they always start with a fresh context containing only the task
+            prompt and workspace context (files from gptme.toml [prompt] files,
+            and context_cmd output when context_mode="full"). This option
+            sanitizes that inherited workspace context.
+
```
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/api.py:127]: Fixed in 4bd988a: added a `logger.warning` at the top of `subagent_reply` when `redact_secrets=True` is passed for subprocess or ACP mode — callers now get a runtime signal rather than a silent no-op.

File: gptme/tools/subagent/context.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py]: Fixed in 72989fe: removed  entirely (the regex was already superseded by  which is now the active colon-style redaction path).

File: gptme/tools/subagent/context.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py:120]: Fixed in 4bd988a: now preserves the original line ending by using  + reassembling with the detected newline.

File: gptme/tools/subagent/api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/api.py:127]: Addressed in 4bd988a: added a warning when redact_secrets=True is requested in non-thread mode (subprocess/ACP) so it's explicit rather than silent. Full support for those modes is a separate follow-up.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

File: gptme/tools/subagent/context.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py]: Fixed in 72989fef7: `_SECRET_LINE_RE` was removed (it was the dead pattern) and replaced by `_COLON_ASSIGN_RE` which is actively used in `_redact_line` (lines 111-118 of context.py). The YAML/colon-style redaction now works correctly.

File: gptme/tools/subagent/context.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/context.py:120]: Fixed in 4bd988a57: `_redact_line` now preserves the original line ending — it captures `ending = "\n" if line.endswith("\n") else ""` and uses `{...}{ending}` in the reconstructed string. Last-line-without-newline case is handled correctly.

File: gptme/tools/subagent/api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TimeToBuildBob) [reply gptme/tools/subagent/api.py:127]: Addressed in two commits: 4bd988a57 adds a warning at the top-level `subagent()` API for subprocess/ACP modes; 3070c6434 (latest) extends this to planner-mode executor threads, adding both the `redact_secrets` forwarding to `_create_subagent_thread` and a warning for subprocess-mode planners. Tests for both paths are in `TestPlannerRedactSecrets`.

**Comment by TimeToBuildBob** [comment]
Rebased onto current master — conflict resolved.

**Status after rebase**: The core `redact_secrets` implementation (context.py + the main changes to subagent/__init__.py, api.py, execution.py) was already merged into master via #2953. What remains in this PR is additional unit test coverage that wasn't in that merge:

- `TestRedactSecretsFromMessages` (5 tests) — `redact_secrets_from_messages()` round-trip tests
- `TestRedactSecretsWarning` — warning emitted when `redact_secrets=True` is used with ACP/subprocess modes
- Additional YAML-style redaction tests (indented keys, multiline, non-secret preservation, trailing-newline fix)

Net diff against master: +109 lines in `tests/test_subagent_unit.py`, no code changes.

CI is running.

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
@greptileai review

**Comment by TimeToBuildBob** [comment]
Fixed the Greptile 4/5 finding in d9b90175f:

**`_run_planner` silently dropped `redact_secrets`**: Added `redact_secrets: bool = False` to `_run_planner`'s signature and wired it through:
- Thread-mode executors now receive `redact_secrets=redact_secrets` when calling `_create_subagent_thread`
- Subprocess-mode executors log a warning (`redact_secrets=True has no effect in subprocess mode`) matching the `api.py` pattern

Also added `TestPlannerRedactSecrets` with two tests covering both paths. All 26 redact_secrets tests pass.
